services/queue/src/utils.js

```diff
@@ -42,86 +41,108 @@ const artifactUtils = {
-        await Promise.all(entries.map((item) => handler.call(item, item)));
-
-        if (!rows.length) {
-          break;
-        } else {
-          const lastRow = rows[rows.length - 1];
-          after_task_id_in = lastRow.task_id;
-          after_run_id_in = lastRow.run_id;
-          after_name_in = lastRow.name;
-        }
+    // Fetch all expired artifacts and batch delete the S3 ones
+    // then remove the entity from the database
+    // repeat until there are no more expired artifacts
+    while (true) {
+      const rows = await db.fns.get_expired_artifacts_for_deletion({
+        expires_in: expires,
+        page_size_in: 1000,
+      });
+      if (!rows.length) {
+        break;
       }
-    };
 
-    await getExpiredArtifacts(expires, async (artifact) => {
-      // Promise that deleted underlying artifact, and keep reference to context
-      let deleted = Promise.resolve();
-
-      // Handle S3 artifacts
-      if (artifact.storageType === 's3') {
-        debug('Deleting expired s3 artifact from bucket: %s, prefix: %s',
-          artifact.details.bucket, artifact.details.prefix);
-        // Delete the right bucket
-        if (artifact.details.bucket === publicBucket.bucket) {
-          deleted = publicBucket.deleteObject(artifact.details.prefix);
-        } else if (artifact.details.bucket === privateBucket.bucket) {
-          deleted = privateBucket.deleteObject(artifact.details.prefix);
-        } else {
-          let err = new Error('Expiring artifact with bucket which isn\'t ' +
-            'configured for use. Please investigate!');
-          err.bucket = artifact.details.bucket;
-          err.taskId = artifact.taskId;
-          err.runId = artifact.runId;
-          monitor.reportError(err);
-          return;
+      const entries = rows.map(exports.artifactUtils.fromDb);
+
+      const s3public = [];
+      const s3private = [];
+
+      for (const entry of entries) {
+        if (entry.storageType === 's3') {
+          if (entry.details.bucket === publicBucket.bucket) {
+            s3public.push(entry);
+          } else if (entry.details.bucket === privateBucket.bucket) {
+            s3private.push(entry);
+          } else {
+            let err = new Error('Expiring artifact with bucket which isn\'t ' +
+              'configured for use. Please investigate!');
+            err.bucket = entry.details.bucket;
+            err.taskId = entry.taskId;
+            err.runId = entry.runId;
+            monitor.reportError(err);
+            continue;
+          }
         }
       }
 
-      // note that 'object' artifacts are deleted at expiration by the object
-      // service, so nothing needs done here.
+      const errors = [];
+      const deleteObjects = async (bucket, entries) => {
+        if (entries.length) {
+          try {
+            const response = await bucket.deleteObjects(entries.map(entry => entry.details.prefix));
+            if (response.Errors && response.Errors.length) {
```

Comment:
Shouldn't this be `Error`? Or did I misread something in those docs? 😄

```suggestion
            if (response.Error && response.Error.length) {
```

Reply:
huh! Controversial one! :) I think the documentation lists response as XML, but when it is parsed by client library, it is being turned into `response.Deleted` and `response.Errors` which are both arrays.. let me try to find some backup for this claim 😬

Reply:
Cannot link the exact section, but here js library explains the response object for this call: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#deleteObjects-property 

```
Errors — (Array<map>) 
```

Reply:
we should switch to typescript 😬